pkg/util/fxutil/provide_comp_test.go

```diff
@@ -451,6 +453,46 @@ func TestFxReturnAnError(t *testing.T) {
 	require.Error(t, app.Err())
 }
 
+func TestFxShutdowner(t *testing.T) {
+	// a bit ugly to save a local reference, but it's okay for test-only code
+	var shutdown compdef.Shutdowner
+	NewAgentComponent := func(reqs requiresShutdownable) provides1 {
+		shutdown = reqs.Shutdown
+		return provides1{First: &firstImpl{}}
+	}
+
+	// define an entry point that waits momentarily then shuts down the application
+	entrypoint := func(first FirstComp) {
+		time.AfterFunc(100*time.Millisecond, func() {
+			shutdown.Shutdown()
+		})
+	}
+
+	// fx-aware upgrade from the regular constructor
+	module := Component(ProvideComponentConstructor(NewAgentComponent))
+
+	// instantiate the application
+	fxApp, _, err := TestApp[FirstComp](
+		fx.Invoke(entrypoint), module,
+	)
+	assert.NoError(t, err)
+
+	// create a wait-loop that will timeout if it takes longer than 1 second
+	ctx, cancel := context.WithCancel(context.Background())
+	time.AfterFunc(1*time.Second, cancel)
+	select {
+	case <-ctx.Done():
+		assert.Fail(t, "waiting for Application timed out, Shutdown() did not work")
+		return
+	case <-fxApp.Done():
+		// if compdef.Shutdown is successfully upgraded to fx.Shutdown,
+		// then this loop should exit here
```

Comment:
What about adding a OnStop lifecyle hook to the test app that when called it flip a boolean value. Then we can also assert that the value is true. 

I'm not sure if that is a bit redundant. What do you think?

Reply:
Tried that approach, but I found that it 1) doesn't work as well and 2) complicates the logic of the test a lot. I like how this test more directly demonstrates exactly what `Shutdown` does.